### ISSUE TITLE
Impossible to surf on 3D printing SE ZIM file

### ISSUE THREAD

**Issue by kelson42** [kiwix/kiwix-android#2303]
Steps:
* Download 3D printing ZIM file
* Load it
* Welcome page displyas properly
* clicking on the first link

I get:
* a blank page

I want:
* The proper page

**Comment by Frans-Lukas** [comment]
Tried on two emulated and one physical device and can't replicate this. All from clean installs though.

**Comment by kelson42** [comment]
Might that be that if we have a ZIM with a service worker, it sill applies?!

**Comment by Frans-Lukas** [comment]
What do you mean? What applies?

**Comment by kelson42** [comment]
This is what I suspected:
* Download http://tmp.kiwix.org/bouquineux.zim and open it and surf a bit
* Then open https://download.kiwix.org/zim/stack_exchange/3dprinting.stackexchange.com_en_all_2020-04.zim and then you have the problem

**Comment by macgills** [comment]
Reproduced
```
2020-08-25 10:10:48.046 W/StaticLayout: maxLineHeight should not be -1.  maxLines:1 lineCount:1
2020-08-25 10:10:48.094 E/chromium: [ERROR:page_load_metrics_update_dispatcher.cc(165)] Invalid first_paint 2.09 s for first_image_paint 1.741 s
2020-08-25 10:10:48.095 D/ZimFileReader: getting mimetype for https://kiwix.app/favicon.ico = image/x-icon
2020-08-25 10:10:48.099 E/kiwix: Unable to get content for url: favicon.ico
2020-08-25 10:10:48.099 E/kiwix: std::exception
2020-08-25 10:10:48.099 D/ZimFileReader: reading  favicon.ico(mime: , size: 0) finished.
2020-08-25 10:10:48.441 D/ZimFileReader: getting mimetype for https://kiwix.app/-/static/fonts/HTML-CSS/TeX/woff/MathJax_Main-Regular.woff?V=2.7.1 = application/font-woff
2020-08-25 10:10:48.465 D/ZimFileReader: reading  -/static/fonts/HTML-CSS/TeX/woff/MathJax_Main-Regular.woff(mime: application/font-woff, size: 34164) finished.
2020-08-25 10:10:48.474 D/ZimFileReader: getting mimetype for https://kiwix.app/-/static/fonts/HTML-CSS/TeX/woff/MathJax_Math-Italic.woff?V=2.7.1 = application/font-woff
2020-08-25 10:10:48.476 D/ZimFileReader: getting mimetype for https://kiwix.app/-/static/fonts/HTML-CSS/TeX/woff/MathJax_Size1-Regular.woff?V=2.7.1 = application/font-woff
2020-08-25 10:10:48.478 D/ZimFileReader: reading  -/static/fonts/HTML-CSS/TeX/woff/MathJax_Math-Italic.woff(mime: application/font-woff, size: 19356) finished.
2020-08-25 10:10:48.486 D/ZimFileReader: reading  -/static/fonts/HTML-CSS/TeX/woff/MathJax_Size1-Regular.woff(mime: application/font-woff, size: 5792) finished.
2020-08-25 10:10:48.775 D/ZimFileReader: getting mimetype for https://kiwix.app/A/topFrame.html = text/html
2020-08-25 10:10:48.785 E/kiwix: Unable to get content for url: A/topFrame.html
2020-08-25 10:10:48.785 E/kiwix: std::exception
2020-08-25 10:10:48.785 D/ZimFileReader: reading  A/topFrame.html(mime: , size: 0) finished.
2020-08-25 10:10:48.835 E/chromium: [ERROR:page_load_metrics_update_dispatcher.cc(165)] Invalid first_paint 2.09 s for first_image_paint 1.741 s
2020-08-25 10:10:48.852 D/kiwix: Loaded URL: https://kiwix.app/A/question/11.html
```

Deeply deeply weird. It loads the 1st page of 3D printing but not the subsequent pages. Moreover I don't see how we intuit that a load is "for a service worker" or not to activate/deactivate at a whim.

The ServiceWorker has seemingly taken over all loading once it loads once but is fine with the index?

It loads pages but not videos on a TED zim.

I could reinitialise the service worker on every zim open? maybe that would work

**Comment by kelson42** [comment]
> I could reinitialise the service worker on every zim open? maybe that would work

We have to that anyway IMO to secure no side effects.

Remark: we need as well to check the behaviours with multiple tabs. Not sure exactly at which level this is attached.

**Comment by macgills** [comment]
tried multiple tabs, once a service worker has been used to load something then it poisons the well. Really puts a damper on multizim tabs.

Enabling ServiceWorkers is like so
```
 if (WebViewFeature.isFeatureSupported(WebViewFeature.SERVICE_WORKER_BASIC_USAGE)) {
      ServiceWorkerControllerCompat.getInstance()
        .setServiceWorkerClient(object : ServiceWorkerClientCompat() {
          override fun shouldInterceptRequest(request: WebResourceRequest): WebResourceResponse? =
            zimReaderContainer.load(request.url.toString())
        })
    }
```
so just setting a client on a singelton instance, we do this at app startup currently.

I can run more experiments after I clear this day's notifications

**Comment by macgills** [comment]
I think this is a problem with `boquineux`? with the `war2c` zim file this issue is not reproducible. I don't really know what is going on though to be perfectly honest

**Comment by kelson42** [comment]
@macgills @rgaudin I believe we have still a problem here. This is a blocker for 3.4 release.

**Comment by rgaudin** [comment]
OK Sorry I'm not subscribed to this repo so just got notified.

@macgills you could test with [this super small zim](http://tmp.kiwix.org/myisago.zim) which is a warc2zim one as well. Probably easier. Just tested with it and could replicate the issue.

The problem I think is quite clear: warc2zim registers a SW for `https://kiwix.app/A/` and this is never unregistered. Manually unregistering from the webview (using chrome dev tools) fixes it so this is probably the way to go.

What's unclear is:
* Can we reset/unregister the SW completely on ZIM opening?
* How individual are the tabs webview? I see the SW thing above as a singleton… but each tab still uses the same base URI, right?

**Comment by macgills** [comment]
1. I see no mechanism android side that would allow me to unregister the SW at all
1.all links have identical base urls, separate zim files are not distinguishable by urls currently. I was working on some sort of system to support that so we could have multiple zim files open concurrently

The api I have is very limited https://developer.android.com/reference/android/webkit/ServiceWorkerController
https://developer.android.com/reference/android/webkit/ServiceWorkerWebSettings

Unless this is a cache mode thing? I still don't know much at all about service workers

**Comment by rgaudin** [comment]
From what I've seen in browsers, it seems to always be separate from cache management.
If there's nothing else, we may inject the unregistration JS? Would it be possible ?

```js
navigator.serviceWorker.getRegistrations().then(function(registrations) {
 for(let registration of registrations) {
  registration.unregister()
} })
```

[SO answer](https://stackoverflow.com/questions/33704791/how-do-i-uninstall-a-service-worker)

**Comment by macgills** [comment]
Yeah it is possible to inject and execute some JS. Very ugly of course but that is the webview for you. I'll look into if that works... probably tomorrow when I am caught up on reviews. Thanks for the help

**Comment by macgills** [comment]
@rgaudin that didn't work 

```
2020-09-11 11:33:53.423 I/chromium: [INFO:CONSOLE(1)] "Uncaught TypeError: Cannot read property 'getRegistrations' of undefined", source:  (1)
```

```
const val UNREGISTER_SERVICE_WORKER_JS =
  """navigator.serviceWorker.getRegistrations().then(function(registrations) {
 for(let registration of registrations) {
  registration.unregister()
} })

kiwixWebView.evaluateJavascript(UNREGISTER_SERVICE_WORKER_JS, null);
"""

**Comment by kelson42** [comment]
@mossroy @Jaifroid Maybe here you can share a hint to help @macgills ?

**Comment by macgills** [comment]
I was successful unregistering via Chrome DevTools :
![image](https://user-images.githubusercontent.com/3358282/92919033-e3255d00-f427-11ea-8fa6-74cd1e5c827f.png)

**Comment by Jaifroid** [comment]
Hmm. The only code I know for unregistering service workers is the code already given above. Is it being injected rcorrectly into a document that is loaded in the WebView? It would be good to look at the code that originally registered the unwanted SW.

As an alternative, is it possible to destroy the WebView and all associated context between ZIM loads? Are you storing any settings you need to keep in JS?

**Comment by macgills** [comment]
Ah I was mistakenly calling the unregister at the wrong time, I was thinking in the old navigation paradigm, apologies for the stupidity.

Unfortunately it still doesn't work but the javascript evaluation at least outputs this?
```
2020-09-15 11:54:17.097 E/ServiceWorker: {}
```
So it did it? Or evaluated to nothing?

**Comment by macgills** [comment]
@rgaudin @Jaifroid  could you supply me with some javascript to debug what is happening? prints should show up in my logs. I am fumbling in the dark a bit

**Comment by Jaifroid** [comment]
@macgills First thing I would do is to add a `.catch(function(err) { console.log(err); });` function after the end of the .then function block - assuming you get console.log() readouts when you eval the JS expression. Additionally you could add `console.log(registration);` before `registration.unregister();`.

If you don't get console.log maybe the webview will show an alert? So in that case you could do `alert('Attempting to unregister Service Worker ` + registtration.index);`, and at least you would know it was attempting to unregister.

If you are unsure whether console.log/alert will show, then add one `console.log('Commencing deregistration');`  (or equivalent alert) before the `Promise.then()` function. If you get that message, but never get the second message, then something is wrong, and the catch function should be thrown.

**Comment by Jaifroid** [comment]
(EDITED typo in one of above code blocks).

**Comment by macgills** [comment]
@rgaudin pushed up a branch here, I think the promises aren't getting enough time to execute? I saw 1 alert when trying to test it. Added some additional logs from the SO question

**Comment by rgaudin** [comment]
thanks, taking a look at it

**Comment by rgaudin** [comment]
Sorry about the long delay, this thing is a bit of a nightmare and I have not been able to really make it work but I think I am bringing enough so that you may.

First of all, I found a bug in another place.
in `CoreWebViewClient`, it's not possible to redirect (actually load) to articles that are not in the ZIM:
in `shouldOverrideUrlLoading()`, `zimReaderContainer.isRedirect(url)` is always true if the url entry is not in the ZIM.
So we then load the `zimReaderContainer.getRedirect(url)` which is blank, resulting in loading `https://kiwix.app/` instead.
I've seen this behavior in early tests of warc2zim and it's annoying. That should be fixed IMO. Also this is somewhat important for what follows.

First, an eyebrow raising fact: while the sotoki ZIM is affected by the remaining SW, it's not the case for most ZIMs, like mwoffliner-made ones. I couldn't find the reason for this and that's a problem to understand the actual problem.

Now, let's define our problem as:

- When using a warc2zim-made ZIM, it registers a ServiceWorker to `https://kiwix.app/A/`
- Webview is thus using it to process any requests to `/A/*`
- SW is never removed
- SW are not part of browser cache (webview or browser) and thus don't get cleared with `webview.clearCaches()`
- SW tries to update themselves after 24h inactivity
- So any regular ZIM opened after a SW-based one still has the SW attached and active on `/A/`, potentially cannibalizing requests.

Obvious way to solve it is thus (as @macgills started), to unregister the SW on leaving the ZIM, so  it's not present on other ZIM and gets re-registered when re-running SW-enabled ZIMs.

The problem here is that the un-registration of a SW is not instantaneous. Unregistering a SW basically tells the browser to not use it anymore for new clients. But existing clients are still using it… untill they move away from its scope.

So to unregister the SW, we need to:
- call `unregister()`
- navigate the webview(s) outside `/A/`

I've achieved to do it **semi-automatically**:
- fix the bug above
- respond to a specific non-`/A` url that the webview is happy with (not much choice actually, so many restrictions). I chose to respond with HTML on `/__exit` for my tests
- call `unregister()` and change location in JS (devtools)
- go to the Library (without reloading the zim obviously)

So this works and that's very good news.

The problem that remains is how to trigger that. @macgills's test attached unregistration code to `onDestroyView()` of the `CoreReaderFragment` but this is too late to perform the JS code and loading that is necessary because the view is already somewhat gone.
I've played with `SystemClock.sleep()` and moving stuff around but that didn't make it.

So all that's left is to trigger that at the right location. Maybe when the `navigation_library` is clicked while in ZIM reading mode… but it's currently handled by the graph… Well, it's now an android problem and I'm sure @macgills will find the right way to do it.

Here's the unregistration code I use:

```js
console.log("tear down");
  function do_unregister() {
    if (!navigator.serviceWorker) {
      return;
    }
    navigator.serviceWorker.getRegistrations().then(async function (registrations) {
      if (registrations.length) {
        console.debug('we do have ' + registrations.length + ' registration(s)');
        var registration = registrations[0];
        registration.unregister().then(function (success) { alert ("SUCC: " + success);}).catch(function (e) {alert("ERR:" + e)});
        window.location.assign("https://kiwix.app/__exit");
      }
    });
  }
  do_unregister();
```

Here's what I serve on `/A/__exit` (hard-coded the response on `load()` of `ZimReaderContainer`:

```html
<html>
<head>
    <title>...</title>
    <script type="text/javascript">console.log("** INSIDE BLANK **");</script>
</head>
<h1>---</h1>
</html>
```

Note that for some reason, sometimes there's a delay (~8s) before it goes away.

I can push my code in some branch but it's full of the remains of the many various attempts :) I need to make sure this is not the result of some side-effect. Will post that later.

**Comment by rgaudin** [comment]
FYI this timeout I sometimes get seems to be related to some deferred loading that can't be resolved and has to timeout.

```
failed to load SourceMap: Could not load content for https://kiwix.app/A/mp_/https:/maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css.map: Connection error: net::ERR_CONNECTION_TIMED_OUT
```

No idea if `webview.stopLoading()` would help with that.

So I think the overall strategy should be something like:

**on opening a ZIM**
- evaluate a JS block that would check if there is/are SW registered
- if not (most of the time), go straight to ZIM
- if there is, launch unregister code, then navigate away to blank page
- once blank page has loaded, load the ZIM homepage

The *blank* page could display a loading indicator or something if that proves to be a second long.
Note that this only happens if we have a SW in the `window` object which would evaluate instantly and thus not disturb the regular use case if not using SW-enabled ZIM files.

**on closing a ZIM** (on moving away from ZIM reading webview actually)
- evaluate a JS block that checks if there's a SW registered
- if not, continue with action
- if there is:
  - stop loading in webview
  - evaluate unregistration code
  - navigate away to *blank* page
  - once blank page loaded, continue with action

Why would we need both? I think we want to optimize loading of ZIM and thus do this cleanup as much as possible when leaving a SW-enabled ZIM. But if the user exits the app directly, this teardown is not done, so we should still be able to do it on startup.

**Comment by macgills** [comment]
**on opening a ZIM**
This didn't work previously, a fresh webview would just say
`2020-09-11 11:33:53.423 I/chromium: [INFO:CONSOLE(1)] "Uncaught TypeError: Cannot read property 'getRegistrations' of undefined", source:  (1)`

I am deeply sceptical we can make this work at all

**Comment by rgaudin** [comment]
If you don't believe it can work, it surely won't 😉

Not sure what's the state of a *fresh webview* but from the 3dprinting webview, I can clearly see the registration:

```js
navigator.serviceWorker.getRegistrations().then(async function (registrations) { console.log(registrations) });
Promise {<pending>}
VM95:1 [ServiceWorkerRegistration]
```

**Comment by macgills** [comment]
is that chrome dev tools? Because I was getting nothing in app

**Comment by rgaudin** [comment]
Yes, it's the devtools. Haven't tried logcat. I suppose we need to wait for initialization (maybe `load`?)

**Comment by macgills** [comment]
The problem I see is if we are deferring the unregistering of previous SWs then we have a likely chance of preforming the unregistering right after we have just opened a new war2czim and registered a SW, seems inappropriate.

**Comment by rgaudin** [comment]
yes, ideally it should:
- know if there's a chance of having a SW (see bellow)
- if not, go straight to ZIM
- if yes, load a placeholder document that would hold the unregistration (or just serve as an environment for it to run)
- once done, load the ZIM content

Now difficulty is knowing whether we're exposed to SW at all or not. Basically that's whether we've last opened a SW-enabled ZIM. We could rely on some ZIM metadata but that would be both tedious and unreliable as a creator could just not set it.
Then we're left with auto-detection. For that we'd need to open the ZIM in a webview wait for it to load and run the detection code.
Doing this transparently seems fragile (what happens if zim is closed before finishing loading?)
We could run that in a hidden webview as part of the ZIM integration to library so it's invisible to the user but that leaves open the question of existing library (or we consider those free of SW zims).
In any of those ways, it means storing in the app (or the local library?) the information of the last used ZIM and whether it's clear of SW or not (either did not have one or have been unregistered).

All of this would be to be able to skip this nasty un-registration procedure so that the 99% of use case are not affected with the special-case, most-likely seconds-consuming of removing a SW.

I know that it sounds like a ton of work to handle a new use case but since we can't wipe the browser's SW, I see no easier way.

**Comment by macgills** [comment]
@kelson42 this would be a significant blocker to release, my current illness means my time is being further reduced on the project so a delayed release of even this friday is probably unlikely. 

Postponing support for Service Workers, reverting all our changes and making a service worker branch of kiwix android to work on compatibility with this feature seems the quickest way forward to me.

**Comment by rgaudin** [comment]
Sounds reasonable to me as well. Changes are quite important and would require some good testing.

**Comment by macgills** [comment]
@kelson42 are you happy to postpone SW support to the 3.5 milestone?

**Comment by macgills** [comment]
It is clumsy and ugly but I am pushing up a draft PR that launches a page on the open of a zim that performs unregistration and calls back to our code to proceed to mainpage, from what I can tell from minor testing it works with that small war2c file and 3d printing stack exchange. Would not be possible without the heroic efforts of @rgaudin

**Comment by macgills** [comment]
I am encountering issues opening new tabs and with the opening of epub/pdfs from bouquineux. @kelson42 can you refresh my memory on what standard zim file contains epubs/pdfs? Just so I can validate our existing functionality works. It seems likely I will have to extract our common functionality around overriding URL loading to be shared among normal loads and SW loads, there are some challenges

**Comment by macgills** [comment]
### Standard zim 
HTML
```
<a href="question/11.html">How do I give 3D-printed parts in PLA a shiny smooth finish?</a>
```
long clicking to open in a new tab uses
```
https://kiwix.app/A/question/11.html
```
normal clicking to open in the same tab uses
```
https://kiwix.app/A/question/11.html
```

### War2c zim
HTML
```
<a href="?ebooks=66&amp;Saint-Exupéry" class="auteur1">Saint-Exupéry</a>
```
long clicking to open in a new tab uses
```
https://kiwix.app/A/mp_/https://www.bouquineux.com/?ebooks=66&Saint-Exup%C3%A9ry
```
normal clicking to open in the same tab uses
```
https://kiwix.app/H/www.bouquineux.com%2F%3Febooks%3D66%26Saint-Exup%25C3%25A9ry
```

@rgaudin @kelson42 any ideas on why the long click url is so bizarre?

**Comment by rgaudin** [comment]
Pinging @ikreymer

**Comment by ikreymer** [comment]
I'm not sure what long clicking is (loading without sw?) but these are the raw URLs loaded from the ZIM that the service worker would load.

I have a suggestion in general. I think this issue could perhaps be resolved more easily by using unique paths for each ZIM file, similar to what kiwix-serve does.

Instead of loading from `https://kiwix.app/`, could kiwix-android load from the `https://kiwix.app/<zimfile>/` URL for each ZIM?

This would avoid having to unregister the service worker and the difficulties associated with that.
I assume the filename is known, or even could be a randomly generated id, since it is not visible to the user, as long as it is unique.

**Comment by rgaudin** [comment]
@ikreymer this was an early suggestion of course that's been ruled out by the way the codebase works. @macgills investigated that and said it's not really doable at the moment.

**Comment by kelson42** [comment]
@macgills If there is an agreement over this ticket's participants that the zimfile id in the URL would help, would you be able please to open a ticket dedicated to that and explain the why? how? etc?

**Comment by macgills** [comment]
It wasn't doable easily, if I implement that I then also basically implement multiple zims open simultaneously, which would be great, just very costly in time which I am in short supply of. It would be a lot better but these issues would still persist

**Comment by kelson42** [comment]
@macgills ah ok, then this is the multzim ticket.

**Comment by macgills** [comment]
>I'm not sure what long clicking is

if the user long presses on a link on their mobile device it triggers this code in the webview

```
  override fun performLongClick(): Boolean {
    val result = hitTestResult
    if (result.type == HitTestResult.SRC_ANCHOR_TYPE) {
      callback.webViewLongClick(result.extra)
      return true
    }
    return super.performLongClick()
  }
```
where `result.extra` is a string containing the url of the content that has been pressed
Next we show a dialog asking if the user wants to open this link in a new tab, they click yes, hey presto new tab with hopefully correctly loaded content.

However trying to load `https://kiwix.app/A/mp_/https://www.bouquineux.com/?ebooks=66&Saint-Exup%C3%A9ry` gets me nothing. Is there any relation between 
```
A/mp_/https://www.bouquineux.com/
```
vs
```
H/www.bouquineux.com
```
Is it a redirect? Or is it absolute nonsense with respect to the zim contents

**Comment by macgills** [comment]
I think opening epubs/pdfs from a zim in an external application was an unwitting sacrifice when we moved away from the content provider. So I can compartmentalise that away as a separate issue for the time being. Should be noted that in a non SW zim that has epubs ie Project Gutenberg, that even with multiple epub apps installed it is not resolving the intent such that it even attempts to open another app. It may be that I am doing something in error so if anyone who has a clearer memory than me remembers opening epubs/pdfs from the app please attempt to do so.

I don't know how to progress the long click to open a new tab issue, we could accept it as a known issue but that doesn't exactly spark joy.

**Comment by kelson42** [comment]
@macgills Yes, please open a dedicated ticket for EPUP/PDF with all the necessary details. For me, it was working after #121 was fixed... so it seems this is not that easy. Put it in 3.5 milestone.

I propose as well to open a ticket for the long click with SW ZIM files and postpone its fix to 3.5 milestone.
